Streaming contribution and fast contribution i/o

porcuquine opened the pull request:
This PR adds streaming contribution to phase2 and makes it the default. This caps memory usage. The somewhat arbitrarily chosen batch size seems to keep memory usage around 20MB without a performance penalty.

This PR also add support for 'raw i/o'. This means that the read/write operations in the streaming hot path avoid expensive transformations in and out of montgomery space. This makes contribution ~10x faster.

The expected net result of this PR is that contribution to even our large circuits should be possible in reasonable time and without extremely beefy hardware. (Actual times and RAM usage will need to be measured on real hardware, of course.)


### Summary of benchmarks for contribution:

Before (non-streaming non-raw) contribution time: 51.62s
Before mem: 53460992 bytes

After (streaming raw) contribution time:  5.26s
After mem: 19374080

Note: these benchmarks don't reflect realistic RAM reduction for large circuits, because the small circuit being used only requires ~50MB RAM. The benefits of streaming will rise for large circuits, which otherwise currently require 55GB RAM. NOTE: streaming is not yet tested on large params, so the claimed cap is a reasoned expectation which still needs verification.

```
(base) ➜  rust-proofs git:(feat/streaming-contribution) ✗ /usr/bin/time -l cargo run --release --bin phase2 contribute-non-streaming winning_poseidon_2kib_b3cfbeb_1_small
a    Finished release [optimized] target(s) in 0.22s
     Running `target/release/phase2 contribute-non-streaming winning_poseidon_2kib_b3cfbeb_1_small`
[filecoin-proofs/src/bin/phase2.rs:687] "in contribute_to_params" = "in contribute_to_params"
03:26:46 [ INFO] contributing to params for circuit: Winning-Poseidon-2KiB-b3cfbeb-Small 1->2
✔ Please randomly press your keyboard (press Return/Enter when finished) · ********
03:26:47 [ INFO] reading small 'before' params as small: winning_poseidon_2kib_b3cfbeb_1_small
03:27:33 [ INFO] phase2::MPCSmall::read() read vector lengths: h=131071, l=91674, contributions=1
03:27:33 [ INFO] successfully read 'before' params, dt_read=46s
03:27:33 [ INFO] making contribution
03:27:33 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
03:27:36 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
03:27:36 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
03:27:38 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
03:27:38 [ INFO] successfully made contribution: 85b1e53599103a7aeb3aec92dcbfa73bf95e626700cb90e97e9ec738f9c1388628afe26e84023aac74055959b5b980ee440cda0ba737b8219544612c9ec05edd, dt_contribute=4s
03:27:38 [ INFO] writing small 'after' params to file: winning_poseidon_2kib_b3cfbeb_2_small
03:27:38 [ INFO] writing contribution hash to file: winning_poseidon_2kib_b3cfbeb_2_small.contrib
03:27:38 [ INFO] successfully made contribution, dt_total=50s
       51.62 real       117.36 user         0.32 sys
  53460992  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     22487  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         1  signals received
        36  voluntary context switches
     50437  involuntary context switches
```

After:
```
(base) ➜  rust-proofs git:(feat/streaming-contribution) ✗ /usr/bin/time -l cargo run --release --bin phase2 contribute winning_poseidon_2kib_b3cfbeb_1_small_raw
    Finished release [optimized] target(s) in 0.22s
     Running `target/release/phase2 contribute winning_poseidon_2kib_b3cfbeb_1_small_raw`
04:13:06 [ INFO] contributing to params for circuit: Winning-Poseidon-2KiB-b3cfbeb-Small 1->2
✔ Please randomly press your keyboard (press Return/Enter when finished) · ********
04:13:06 [ INFO] making contribution
04:13:06 [ INFO] making streamer from small 'before' params: winning_poseidon_2kib_b3cfbeb_1_small_raw
04:13:06 [ INFO] writing small 'after' params to file: winning_poseidon_2kib_b3cfbeb_2_small_raw
04:13:06 [ INFO] phase2::MPCParameters::contribute() beginning streaming h
04:13:06 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:07 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:08 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of h
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished streaming h
04:13:09 [ INFO] phase2::MPCParameters::contribute() beginning streaming l
04:13:09 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:09 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:09 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:10 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() finished batch_exp of l
04:13:11 [ INFO] phase2::MPCParameters::contribute() finished streaming l
04:13:11 [ INFO] successfully made contribution: 77c8f42b050bfabe26c18efb361ae52d0f99f4bd52aac01539e1825e723daf6618f7ffd8e899c94e1b9fe81c604862a4357d6c4211da273a1a5beab0a1e01839, dt_contribute=4s
04:13:11 [ INFO] writing contribution hash to file: winning_poseidon_2kib_b3cfbeb_2_small_raw.contrib
04:13:11 [ INFO] successfully made contribution, dt_total=4s
        5.26 real        70.33 user         0.21 sys
  19374080  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     16551  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         1  signals received
         7  voluntary context switches
     26203  involuntary context switches```

DrPeterVanNostrand commented:
Whoo, those are nice benchmarks.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -1,75 +1,71 @@
 use storage_proofs::porep::stacked::{StackedCircuit, StackedCompound, StackedDrg};
 use storage_proofs::post::fallback::{FallbackPoSt, FallbackPoStCircuit, FallbackPoStCompound};
 
+const CHUNK_SIZE: usize = 10_000;
```
Was this value chosen because it's roughly the read buffer size divided by the byte length of a serialized G1Affine?

porcuquine replied:
The buffer size used by the `Streamer` is chosen so one chunk will fit into the buffer. This`CHUNK_SIZE` was chosen to be 'good enough': it keeps memory usage low without an observable performance penalty. It's conceivable that a different strategy for all of this might provide some micro-optimization, but this is what emerged heuristically and has been tested.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -79,98 +75,201 @@ enum Hasher {
+        other => panic!("invalid param-size in params filename: {}", other),
+    };
+
+    let raw_fmt = if split.len() > 6 && split[6] == "raw" {
```
```rust
let raw_fmt = split.get(6) == Some(&"raw");
```

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+
+    let contrib = streamer
+        .contribute(&mut rng, file_after, CHUNK_SIZE)
+        .expect("failed to make contribution");
```
Could you please add the IO error to the panic message

```rust
let contrib = streamer
    .contribute(...)
    .unwrap_or_else(|e| {
        panic!("failed to make streaming contribution: {}", e);
    });
```

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
-    let start_total = Instant::now();
+fn convert_small<'a>(path_before: &'a str) {
+    let (proof, hasher, sector_size, head, param_number, param_size, read_raw) =
+        parse_params_filename(path_before);
```
Could you add an assertion

```rust
# TODO: change this if we update the large MPC params (and G2Affine) to support the raw serialization format.
assert!(param_size.is_small(), "converting large params to raw format is not currently supported");
```

DrPeterVanNostrand replied:
And maybe the check

```rust
if read_raw {
    info!("params are already in the raw format, exiting early");
    return;
}
```

porcuquine replied:
I think your proposed assertion is already covered a few lines below:

```
    let mut streamer = if param_size.is_large() {
        panic!("cannot convert large param format");
```

If we *do* end up supporting raw format for large params, I assume we would not add it to the `convert_small` function anyway.

porcuquine replied:
Regarding the proposed check and early exit: this function actually converts both to and from raw. When called on params in raw format, it will convert to 'non-raw'.

DrPeterVanNostrand replied:
> [porcuquine]: I think your proposed assertion is already covered a few lines below
> [drpetervannostrant]: this assertion will remove a panic when streamer is created.
Yea, I just wanted to exit before the "converting params ..." log.

> this function actually converts both to and from raw
Sorry, I didn't realize that. Please disregard.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
-        "the number of contributions must be one less than the number of parameter files"
+    let write_raw = !read_raw;
+
+    info!(
```
Could you please update this log

```rust
info!(
    "copying params into raw format for circuit: {proof}-{hasher}-{sector_size}-{head}-{num}-{param_size}",
    proof = proof.pretty_print(),
    hasher = hasher.pretty_print(),
    sector_size = sector_size.pretty_print(),
    head = head,
    num = param_number,
    param_size = param_size.pretty_print()
);
```

porcuquine replied:
Done, but with modification to indicate whether we are converting to or from raw.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
-            }
-        };
+    info!(
+        "making streamer from small 'before' params: {}",
```
The phase2-cli uses "before" to indicate params prior to a contribution is made, could you update this log to:

```rust
info!("creating streamer from non-raw params: {}", path_before);
```

porcuquine replied:
Done, but correctly indicating whether the source is raw or non-raw.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
-        info!(
-            "successfully verified param transition, dt_verify={}s",
-            start_verification.elapsed().as_secs()
+    info!("writing small 'after' params to file: {}", path_after);
```
"after" means params after a contribution is made, could you change this log to something like:

```rust
info!("writing raw formatted params to file: {}", path_after);
```

porcuquine replied:
Done, but correctly indicating whether raw or non-raw params are being written.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+    let (proof, hasher, sector_size, head, prev_param_number, param_size, read_raw) =
+        parse_params_filename(path_before);
+
+    dbg!("in contribute_to_params");
```
Delete

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+        parse_params_filename(path_before);
+
+    dbg!("in contribute_to_params");
+    assert_eq!(false, read_raw, "Raw format only supported when streaming.");
```
This function is the non-streaming contribution, right? `MPCSmall::contribute()` can operate on both raw and non-raw?

porcuquine replied:
Yes. This assertion and the FIXME both date to an earlier stage in development when it was not yet possible to create an `MPCSmall` from raw format. I have removed the assertion and addressed the FIXME so this now-possible path is allowed.

DrPeterVanNostrand commented on filecoin-proofs/src/bin/phase2.rs:
```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+        info!("reading small 'before' params as small: {}", path_before);
+        let file_before = File::open(path_before).unwrap();
+        let mut reader = BufReader::with_capacity(1024 * 1024, file_before);
+        let read_raw_affine = false; // FIXME: need to support the true path also.
```
We can just pass `false` in as an argument.

porcuquine replied:
Since it's now okay to create an `MPCSmall` from raw params, I have instead removed this line and pass the value of `read_raw` received from parsing the input file. This resolves the FIXME.

DrPeterVanNostrand replied:
Nice, thanks.

DrPeterVanNostrand commented:
Just a few comments, mostly about logs.

porcuquine commented:
@DrPeterVanNostrand I think I have addressed all of your comments in my latest commit. In some cases, the resolution was different than your suggestion, for reasons I have clarified in my responses above.

DrPeterVanNostrand commented on filecoin-proofs/Cargo.toml:
```diff
@@ -48,8 +48,8 @@ sha2 = { version = "0.8.3", package = "sha2ni" }
-phase2 = { package = "phase21", version = "0.6.0" }
-simplelog = "0.8.0"
+phase2 = { version = "0.7.0", package = "phase21" }
+simplelog = "0.7.4"
```
We should accept the `simplelog` version bump that's currently in master.

DrPeterVanNostrand commented:
Everything looks good to me. Thank you for these changes!
